source/delegate/contracts/Delegate.sol

```diff
@@ -0,0 +1,165 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity 0.8.23;
+
+import "./interfaces/IDelegate.sol";
+import "@airswap/swap-erc20/contracts/interfaces/ISwapERC20.sol";
+import { ERC20 } from "solady/src/tokens/ERC20.sol";
+import { Ownable } from "solady/src/auth/Ownable.sol";
+import { SafeTransferLib } from "solady/src/utils/SafeTransferLib.sol";
+
+/**
+ * @title Delegate: Deployable Trading Rules for the AirSwap Network
+ * @notice Supports fungible tokens (ERC-20)
+ * @dev inherits IDelegate, Ownable and uses SafeTransferLib
+ */
+contract Delegate is IDelegate, Ownable {
+  // The Swap contract to be used to settle trades
+  ISwapERC20 public swapERC20;
+
+  // Mapping of sender to senderToken to to signerToken to Rule
+  mapping(address => mapping(address => mapping(address => Rule))) public rules;
+
+  // Mapping of signer to authorized signatory
+  mapping(address => address) public authorized;
+
+  /**
+   * @notice Contract Constructor
+   */
+  constructor(ISwapERC20 _swapERC20) {
+    swapERC20 = _swapERC20;
+  }
+
+  /**
+   * @notice Set a Trading Rule
+   * @param _senderToken address Address of an ERC-20 token the consumer would send
+   * @param _senderAmount uint256 Maximum amount of ERC-20 token the sender wants to swap
+   * @param _signerToken address Address of an ERC-20 token the delegate would recieve
+   * @param _signerAmount uint256 Minimum amount of ERC-20 token the delegate would recieve
+   */
+  function setRule(
+    address _senderWallet,
+    address _senderToken,
+    uint256 _senderAmount,
+    address _signerToken,
+    uint256 _signerAmount
+  ) external {
+    if (authorized[_senderWallet] != address(0)) {
+      if (authorized[_senderWallet] != msg.sender) revert SenderInvalid();
+    } else {
+      if (_senderWallet != msg.sender) revert SenderInvalid();
+    }
+
+    rules[_senderWallet][_senderToken][_signerToken] = Rule(
+      _senderWallet,
+      _senderToken,
+      _senderAmount,
+      _signerToken,
+      _signerAmount
+    );
+
+    emit SetRule(
+      _senderWallet,
+      _senderToken,
+      _senderAmount,
+      _signerToken,
+      _signerAmount
+    );
+  }
+
+  /**
+   * @notice Unset a Trading Rule
+   * @dev only callable by the owner of the contract, removes from a mapping
+   * @param _senderToken address Address of an ERC-20 token the sender would send
+   * @param _signerToken address Address of an ERC-20 token the delegate would receive
+   */
+  function unsetRule(
+    address _senderWallet,
+    address _senderToken,
+    address _signerToken
+  ) external {
+    if (authorized[_senderWallet] != address(0)) {
+      if (authorized[_senderWallet] != msg.sender) revert SenderInvalid();
+    } else {
+      if (_senderWallet != msg.sender) revert SenderInvalid();
+    }
+    Rule storage rule = rules[_senderWallet][_senderToken][_signerToken];
+    rule.senderAmount = 0;
```

Comment:
Would we gain anything by zeroing out the other fields?

Reply:
I originally didn't do it to save on gas. But the additional security might be worth the gas (which won't be much)

Reply:
How about just something like `delete rules[_senderWallet][_senderToken][_signerToken];`

Reply:
Absolutely... No idea why I wrote that horror...